### PR TITLE
EE-1159: lock genesis validator bids for 90 days

### PR DESCRIPTION
**Pull request by henrytill**
Ref: https://casperlabs.atlassian.net/browse/EE-1136 (subtask of this ticket)

This PR converts the existing era-based lock-in cliff to a timestamp-based one.  Genesis validator bids are locked for 90 days.

Subsequent PRs will address the vesting schedule and apply similar logic to delegators (it is currently validator-only).

### CODE REVIEW

File: node/src/components/block_executor.rs [hunk 1]
```diff
@@ -235,13 +235,15 @@ impl BlockExecutor {
                     .iter()
                     .map(|&vid| SlashItem::new(vid))
                     .collect();
+                let era_end_timestamp_millis = state.finalized_block.timestamp().millis();
```
Comment (henrytill) [inline node/src/components/block_executor.rs:238]: cc @afck @Fraser999: please confirm that this is right timestamp to use when calling the step function.

Comment (afck) [reply node/src/components/block_executor.rs:238]: Yes, that's right.

File: grpc/test_support/src/internal/mod.rs [hunk 2]
```diff
@@ -37,7 +37,9 @@ pub use wasm_test_builder::{
 
 pub const DEFAULT_VALIDATOR_SLOTS: u32 = 5;
 pub const DEFAULT_AUCTION_DELAY: u64 = 3;
-pub const DEFAULT_LOCKED_FUNDS_PERIOD: EraId = 15;
+/// Default lock-in period of 90 days
+/// <https://www.wolframalpha.com/input/?i=convert+90+days+to+milliseconds>
+pub const DEFAULT_LOCKED_FUNDS_PERIOD_MILLIS: u64 = 7776000000;
```
Comment (afck) [inline grpc/test_support/src/internal/mod.rs:42]: You could spell out the 90 days as a product:
```suggestion
pub const DEFAULT_LOCKED_FUNDS_PERIOD_MILLIS: u64 = 90 * 24 * 60 * 60 * 1000;
```

File: node/src/components/chainspec_loader/chainspec.rs [hunk 3]
```diff
@@ -209,10 +209,8 @@ pub struct GenesisConfig {
     /// If you bond with a sufficient bid in era N, you will be a validator in era N +
     /// auction_delay + 1
     pub(crate) auction_delay: u64,
-    /// The delay for the payout of funds, in eras. If a withdraw request is included in a block in
-    /// era N (other than the last one), they are paid out in the last block of era N +
-    /// locked_funds_period.
-    pub(crate) locked_funds_period: EraId,
+    /// Number of days (in milliseconds) after genesis that a genesis validator's bid is locked
```
Comment (afck) [inline node/src/components/chainspec_loader/chainspec.rs:212]:
```suggestion
    /// Number of milliseconds after genesis that a genesis validator's bid is locked
```

File: node/src/components/chainspec_loader/chainspec.rs [hunk 4]
```diff
@@ -301,7 +299,7 @@ impl GenesisConfig {
             timestamp,
             validator_slots,
             auction_delay,
-            locked_funds_period,
+            locked_funds_period_millis: locked_funds_period,
```
Comment (afck) [inline node/src/components/chainspec_loader/chainspec.rs:302]: I'd also rename the local variable then.

File: node/src/components/chainspec_loader/config.rs [hunk 5]
```diff
@@ -22,7 +22,11 @@ const DEFAULT_CHAIN_NAME: &str = "casper-devnet";
 const DEFAULT_ACCOUNTS_CSV_PATH: &str = "accounts.csv";
 const DEFAULT_VALIDATOR_SLOTS: u32 = 5;
 const DEFAULT_AUCTION_DELAY: u64 = 3;
-const DEFAULT_LOCKED_FUNDS_PERIOD: EraId = 15;
+
+/// Default lock-in period of 90 days
+/// <https://www.wolframalpha.com/input/?i=convert+90+days+to+milliseconds>
+const DEFAULT_LOCKED_FUNDS_PERIOD_MILLIS: u64 = 7776000000; // 90 days
```
Comment (afck) [inline node/src/components/chainspec_loader/config.rs:28]: As above, you could write it as a product?

File: resources/joiner/chainspec.toml.in [hunk 6]
```diff
@@ -16,9 +16,8 @@ accounts_path = './accounts.csv'
 validator_slots = 5
 # Number of eras before an auction actually defines the set of validators. If you bond with a sufficient bid in era N, you will be a validator in era N + auction_delay + 1
 auction_delay = 3
-# The delay for the payout of funds, in eras. If a withdraw request is included in a block in era N
-# (other than the last one), they are paid out in the last block of era N + locked_funds_period.
-locked_funds_period = 15
+# Number of days (in milliseconds) after genesis that a genesis validator's bid is locked
```
Comment (afck) [inline resources/joiner/chainspec.toml.in:19]:
```suggestion
# Number of milliseconds after genesis that a genesis validator's bid is locked
# 7776000000 ms = 90 days
```
(Also in the other chainspecs.)

File: resources/local/chainspec.toml.in [hunk 7]
```diff
@@ -16,9 +16,8 @@ accounts_path = '${BASEDIR}/resources/local/accounts.csv'
 validator_slots = 5
 # Number of eras before an auction actually defines the set of validators. If you bond with a sufficient bid in era N, you will be a validator in era N + auction_delay + 1
 auction_delay = 3
-# The delay for the payout of funds, in eras. If a withdraw request is included in a block in era N
-# (other than the last one), they are paid out in the last block of era N + locked_funds_period.
-locked_funds_period = 15
+# Number of days (in milliseconds) after genesis that a genesis validator's bid is locked
+locked_funds_period_millis = 7776000000
```
Comment (afck) [inline resources/local/chainspec.toml.in:20]: Maybe for the local one it should just be a few minutes, so it's easier to test?

File: types/src/auction/constants.rs [hunk 8]
```diff
@@ -50,6 +50,9 @@ pub const ARG_AUCTION_DELAY: &str = "auction_delay";
 pub const ARG_LOCKED_FUNDS_PERIOD: &str = "locked_funds_period";
 /// Named constant for `unbonding_delay`
 pub const ARG_UNBONDING_DELAY: &str = "unbonding_delay";
+/// Named constant for `era_end_timestamp_mills`;
```
Comment (afck) [inline types/src/auction/constants.rs:53]:
```suggestion
/// Named constant for `era_end_timestamp_millis`;
```

File: node/src/components/block_executor.rs (hunk 1 above)
Comment (henrytill) [reply node/src/components/block_executor.rs:238]: Cool.

File: grpc/test_support/src/internal/mod.rs (hunk 2 above)
Comment (henrytill) [reply grpc/test_support/src/internal/mod.rs:42]: Good suggestion, will do.

File: node/src/components/chainspec_loader/chainspec.rs (hunk 4 above)
Comment (henrytill) [reply node/src/components/chainspec_loader/chainspec.rs:302]: Whoops, missed that one.

File: grpc/test_support/src/internal/mod.rs (hunk 2 above)
Comment (henrytill) [reply grpc/test_support/src/internal/mod.rs]: Donezo.

File: node/src/components/chainspec_loader/chainspec.rs (hunk 3 above)
Comment (henrytill) [reply node/src/components/chainspec_loader/chainspec.rs]: Donezo.

File: node/src/components/chainspec_loader/chainspec.rs (hunk 4 above)
Comment (henrytill) [reply node/src/components/chainspec_loader/chainspec.rs]: Donezo.

File: node/src/components/chainspec_loader/config.rs (hunk 5 above)
Comment (henrytill) [reply node/src/components/chainspec_loader/config.rs]: Donezo.

File: resources/joiner/chainspec.toml.in (hunk 6 above)
Comment (henrytill) [reply resources/joiner/chainspec.toml.in]: Donezo.

File: resources/local/chainspec.toml.in (hunk 7 above)
Comment (henrytill) [reply resources/local/chainspec.toml.in]: Set to 5 minutes.

File: types/src/auction/constants.rs (hunk 8 above)
Comment (henrytill) [reply types/src/auction/constants.rs]: Donezo.

**Review by afck — APPROVED** [review]
Looks good to me — but I'm not very familiar with the EE code.

**Review by EdHastingsCasperLabs — APPROVED** [review]

File: grpc/server/src/engine_server/mappings/ipc/step.rs [hunk 9]
```diff
@@ -189,6 +189,16 @@ impl TryFrom<StepRequest> for ipc::StepRequest {
...
         result.set_reward_items(reward_items.into());
 
+        let evict_items = {
+            let mut ret: Vec<ipc::EvictItem> = vec![];
+            for item in step_request.evict_items.into_iter() {
+                let ipc = item.try_into()?;
+                ret.push(ipc);
+            }
+            ret
+        };
```
Comment (afck) [inline grpc/server/src/engine_server/mappings/ipc/step.rs:199]: Maybe more concise:
```rust
let evict_items = step_request
    .evict_items
    .into_iter()
    .map(TryInto::try_into)
    .collect::<Result<Vec<_, _>>>()?;
```
What is the vector turned `into()` in the next line? Can we even avoid `Vec` allocation?
(Anyway, none of this is important, of course!)

Comment (henrytill) [reply grpc/server/src/engine_server/mappings/ipc/step.rs:199]: Agreed, but this is not my code...just coming from the merge.

Comment (afck) [reply grpc/server/src/engine_server/mappings/ipc/step.rs]: Right, sorry!

**Pull request merged**